File: python/2015/day21.py
# python day21.py ../../data/2015/day21.dat

import itertools
import sys


def parse_input(input_data):
    stats = {}
    for line in input_data.strip().split("\n"):
        key, value = line.split(": ")
        stats[key] = int(value)
    return stats["Hit Points"], stats["Damage"], stats["Armor"]


def player_wins(boss_hp, boss_def, boss_att, player_def, player_att):
    player_hp = 100
    while player_hp > 0 and boss_hp > 0:
        boss_hp -= max(1, player_att - boss_def)
        if boss_hp > 0:
            player_hp -= max(1, boss_att - player_def)
    return player_hp > 0


def part1(boss_hp, boss_damage, boss_armor):
    weapons = [
        {"cost": 8, "damage": 4, "armor": 0},
        {"cost": 10, "damage": 5, "armor": 0},
        {"cost": 25, "damage": 6, "armor": 0},
        {"cost": 40, "damage": 7, "armor": 0},
        {"cost": 74, "damage": 8, "armor": 0},
    ]
    armor = [
        {"cost": 0, "damage": 0, "armor": 0},
        {"cost": 13, "damage": 0, "armor": 1},
        {"cost": 31, "damage": 0, "armor": 2},
        {"cost": 53, "damage": 0, "armor": 3},
        {"cost": 75, "damage": 0, "armor": 4},
        {"cost": 102, "damage": 0, "armor": 5},
    ]
    rings = [
        {"cost": 0, "damage": 0, "armor": 0},
        {"cost": 25, "damage": 1, "armor": 0},
        {"cost": 50, "damage": 2, "armor": 0},
        {"cost": 100, "damage": 3, "armor": 0},
        {"cost": 20, "damage": 0, "armor": 1},
        {"cost": 40, "damage": 0, "armor": 2},
        {"cost": 80, "damage": 0, "armor": 3},
    ]

    min_cost = float('inf')
    for w in weapons:
        for a in armor:
            for r1, r2 in itertools.combinations(rings, 2):
                if r1["cost"] == 0 and r2["cost"] == 0:
                    continue
                if r1["cost"] == r2["cost"]:
                    continue
                player_att = w["damage"] + r1["damage"] + r2["damage"]
                player_def = a["armor"] + r1["armor"] + r2["armor"]
                cost = w["cost"] + a["cost"] + r1["cost"] + r2["cost"]
                if player_wins(boss_hp, boss_armor, boss_damage, player_def, player_att) and cost < min_cost:
                    min_cost = cost
    return min_cost


def part2(boss_hp, boss_damage, boss_armor):
    weapons = [
        {"cost": 8, "damage": 4, "armor": 0},
        {"cost": 10, "damage": 5, "armor": 0},
        {"cost": 25, "damage": 6, "armor": 0},
        {"cost": 40, "damage": 7, "armor": 0},
        {"cost": 74, "damage": 8, "armor": 0},
    ]
    armor = [
        {"cost": 0, "damage": 0, "armor": 0},
        {"cost": 13, "damage": 0, "armor": 1},
        {"cost": 31, "damage": 0, "armor": 2},
        {"cost": 53, "damage": 0, "armor": 3},
        {"cost": 75, "damage": 0, "armor": 4},
        {"cost": 102, "damage": 0, "armor": 5},
    ]
    rings = [
        {"cost": 0, "damage": 0, "armor": 0},
        {"cost": 25, "damage": 1, "armor": 0},
        {"cost": 50, "damage": 2, "armor": 0},
        {"cost": 100, "damage": 3, "armor": 0},
        {"cost": 20, "damage": 0, "armor": 1},
        {"cost": 40, "damage": 0, "armor": 2},
        {"cost": 80, "damage": 0, "armor": 3},
    ]

    max_cost = 0
    for w in weapons:
        for a in armor:
            for r1, r2 in itertools.combinations(rings, 2):
                if r1["cost"] == 0 and r2["cost"] == 0:
                    continue
                if r1["cost"] == r2["cost"]:
                    continue
                player_att = w["damage"] + r1["damage"] + r2["damage"]
                player_def = a["armor"] + r1["armor"] + r2["armor"]
                cost = w["cost"] + a["cost"] + r1["cost"] + r2["cost"]
                if not player_wins(boss_hp, boss_armor, boss_damage, player_def, player_att) and cost > max_cost:
                    max_cost = cost
    return max_cost


if __name__ == "__main__":
    if len(sys.argv) > 1:
        with open(sys.argv[1], "r") as f:
            input_data = f.read()
    else:
        input_data = ""
    boss_hp, boss_damage, boss_armor = parse_input(input_data)
    print("Day 21 Part 1:", part1(boss_hp, boss_damage, boss_armor))
    print("Day 21 Part 2:", part2(boss_hp, boss_damage, boss_armor))

File: java/2015/Day21.java
import module java.base;

public class Day21 {
    private record Item(String name, int cost, int damage, int armor) {}

    private static final Item[] WEAPONS = {
        new Item("Dagger", 8, 4, 0), new Item("Shortsword", 10, 5, 0),
        new Item("Warhammer", 25, 6, 0), new Item("Longsword", 40, 7, 0),
        new Item("Greataxe", 74, 8, 0)
    };
    private static final Item[] ARMOR = {
        new Item("None", 0, 0, 0), new Item("Leather", 13, 0, 1),
        new Item("Chainmail", 31, 0, 2), new Item("Splintmail", 53, 0, 3),
        new Item("Bandedmail", 75, 0, 4), new Item("Platemail", 102, 0, 5)
    };
    private static final Item[] RINGS = {
        new Item("None", 0, 0, 0), new Item("Damage +1", 25, 1, 0),
        new Item("Damage +2", 50, 2, 0), new Item("Damage +3", 100, 3, 0),
        new Item("Defense +1", 20, 0, 1), new Item("Defense +2", 40, 0, 2),
        new Item("Defense +3", 80, 0, 3)
    };

    public static void main(String[] args) {
        var input = readInput("../../data/2015/day21.dat");
        if (input.isEmpty()) return;

        var bossHp = 0; var bossAtt = 0; var bossDef = 0;
        for (var line : input.lines().toList()) {
            var parts = line.split(": ");
            var val = Integer.parseInt(parts[1]);
            switch (parts[0]) {
                case "Hit Points" -> bossHp = val;
                case "Damage" -> bossAtt = val;
                case "Armor" -> bossDef = val;
            }
        }

        var minCost = Integer.MAX_VALUE;
        var maxCost = 0;

        for (var r1 : RINGS) {
            for (var r2 : RINGS) {
                if (r1.cost() == 0 && r2.cost() == 0) continue;
                if (r1.cost() == r2.cost()) continue;
                for (var w : WEAPONS) {
                    for (var a : ARMOR) {
                        var playerDmg = r1.damage() + r2.damage() + w.damage();
                        var playerDef = r1.armor() + r2.armor() + a.armor();
                        var cost = r1.cost() + r2.cost() + w.cost() + a.cost();
                        if (playerWins(bossHp, bossAtt, bossDef, playerDef, playerDmg)) {
                            minCost = Math.min(minCost, cost);
                        } else {
                            maxCost = Math.max(maxCost, cost);
                        }
                    }
                }
            }
        }

        System.out.println("Part 1: " + minCost);
        System.out.println("Part 2: " + maxCost);
    }

    private static String readInput(String filePath) {
        try {
            return Files.readString(Paths.get(filePath)).trim();
        } catch (NoSuchFileException e) {
            System.out.println("Data file missing: " + filePath);
        } catch (IOException ex) {
        }
        return "";
    }

    private static boolean playerWins(int bossHp, int bossAtt, int bossDef, int playerDef, int playerAtt) {
        var hp = 100;
        while (hp > 0 && bossHp > 0) {
            bossHp -= Math.max(1, playerAtt - bossDef);
            if (bossHp > 0) hp -= Math.max(1, bossAtt - playerDef);
        }
        return hp > 0;
    }
}
